File: apps/plant_shop_c-sharp/plant_shop_c-sharp/Repositories/OrderItemRepository.cs
using Npgsql;
using plant_shop_c_sharp.Models;
using System;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace plant_shop_c_sharp.Repositories
{
    /// <summary>
    /// Repository CRUD pour les items de commande.
    /// </summary>
    public class OrderItemRepository : BaseRepository
    {
        /// <summary>
        /// Constructeur avec injection de la source de donnees.
        /// </summary>
        /// <param name="dataSource">Source Npgsql.</param>
        public OrderItemRepository(NpgsqlDataSource dataSource) : base(dataSource) { }

        /// <summary>
        /// Trouve un item par ID.
        /// </summary>
        /// <param name="id">ID a rechercher.</param>
        /// <returns>OrderItem ou null.</returns>
        public async Task<OrderItem?> FindByIdAsync(int id)
        {
            await using var conn = GetConnection();
            await using var cmd = new NpgsqlCommand("SELECT * FROM order_items WHERE id = @id", conn);
            cmd.Parameters.AddWithValue("id", id);

            await using var reader = await cmd.ExecuteReaderAsync();
            if (await reader.ReadAsync())
            {
                return MapOrderItem(reader);
            }
            return null;
        }

        /// <summary>
        /// Liste les items d une commande.
        /// </summary>
        /// <param name="orderId">ID de la commande.</param>
        /// <returns>Liste des items.</returns>
        public async Task<List<OrderItem>> FindByOrderIdAsync(int orderId)
        {
            var items = new List<OrderItem>();
            await using var conn = GetConnection();
            await using var cmd = new NpgsqlCommand("SELECT * FROM order_items WHERE order_id = @orderId", conn);
            cmd.Parameters.AddWithValue("orderId", orderId);

            await using var reader = await cmd.ExecuteReaderAsync();
            while (await reader.ReadAsync())
            {
                items.Add(MapOrderItem(reader));
            }
            return items;
        }

        /// <summary>
        /// Cree un nouvel item.
        /// </summary>
        /// <param name="item">Item a creer.</param>
        /// <returns>Item avec ID genere.</returns>
        public async Task<OrderItem> CreateAsync(OrderItem item)
        {
            await using var conn = GetConnection();
            var sql = "INSERT INTO order_items (order_id, plant_id, quantity, price) VALUES (@orderId, @plantId, @qty, @price) RETURNING id";
            await using var cmd = new NpgsqlCommand(sql, conn);

            cmd.Parameters.AddWithValue("orderId", item.OrderId);
            cmd.Parameters.AddWithValue("plantId", item.PlantId.HasValue ? item.PlantId.Value : DBNull.Value);
            cmd.Parameters.AddWithValue("qty", item.Quantity);
            cmd.Parameters.AddWithValue("price", item.Price);

            item.Id = (int)await cmd.ExecuteScalarAsync();
            return item;
        }

        /// <summary>
        /// Supprime tous les items d une commande.
        /// </summary>
        /// <param name="orderId">ID de la commande.</param>
        public async Task DeleteByOrderIdAsync(int orderId)
        {
            await using var conn = GetConnection();
            await using var cmd = new NpgsqlCommand("DELETE FROM order_items WHERE order_id = @orderId", conn);
            cmd.Parameters.AddWithValue("orderId", orderId);
            await cmd.ExecuteNonQueryAsync();
        }


        /// <summary>
        /// Mappe un reader vers un OrderItem.
        /// </summary>
        /// <param name="reader">Reader Npgsql.</param>
        /// <returns>Entite OrderItem.</returns>
        private OrderItem MapOrderItem(NpgsqlDataReader reader)
        {
            return new OrderItem
            {
                Id = reader.GetInt32(reader.GetOrdinal("id")),
                OrderId = reader.GetInt32(reader.GetOrdinal("order_id")),
                PlantId = reader.IsDBNull(reader.GetOrdinal("plant_id")) ? null : reader.GetInt32(reader.GetOrdinal("plant_id")),
                Quantity = reader.GetInt32(reader.GetOrdinal("quantity")),
                Price = reader.GetDecimal(reader.GetOrdinal("price"))
            };
        }
    }
}

File: apps/plant_shop_java/Structure_en_micro_services/plant_shop_java_microservices/order-service/src/repository/OrderItemRepository.java
package order.repository;

import order.model.OrderItem;
import order.model.PlantStock;
import java.math.BigDecimal;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;

/**
 * Repository pour les items de commande.
 */
public final class OrderItemRepository {

    private final Connection db;

    /**
	 * Constructeur.
	 * @param db Connexion à la base de données
	 */
	public OrderItemRepository(Connection db) {
        this.db = db;
    }

    /**
	 * Mappe un ResultSet vers un OrderItem.
	 * @param rs ResultSet
	 * @return OrderItem
	 * @throws SQLException En cas d'erreur SQL
	 */
	OrderItem map(ResultSet rs) throws SQLException {
        return new OrderItem(
            rs.getInt("id"),
            rs.getInt("order_id"),
            rs.getInt("plant_id"),
            rs.getInt("quantity"),
            rs.getBigDecimal("price")
        );
    }

    /**
	 * Crée un item de commande.
	 * @param item Item à créer
	 * @return ID généré
	 * @throws SQLException En cas d'erreur SQL
	 */
	public int create(OrderItem item) throws SQLException {
        try (PreparedStatement ps = db.prepareStatement(
            "INSERT INTO order_items(order_id, plant_id, quantity, price) VALUES (?, ?, ?, ?)",
            PreparedStatement.RETURN_GENERATED_KEYS)) {
            ps.setInt(1, item.orderId());
            ps.setInt(2, item.plantId());
            ps.setInt(3, item.quantity());
            ps.setBigDecimal(4, item.price());
            ps.executeUpdate();
            try (ResultSet rs = ps.getGeneratedKeys()) {
                rs.next();
                return rs.getInt(1);
            }
        }
    }

    /**
	 * Trouve les items d'une commande.
	 * @param orderId ID de la commande
	 * @return Liste des items
	 * @throws SQLException En cas d'erreur SQL
	 */
	public List<OrderItem> findByOrder(int orderId) throws SQLException {
        List<OrderItem> out = new ArrayList<>();
        try (PreparedStatement ps = db.prepareStatement("SELECT * FROM order_items WHERE order_id=?")) {
            ps.setInt(1, orderId);
            try (ResultSet rs = ps.executeQuery()) {
                while (rs.next()) {
                    out.add(map(rs));
                }
            }
        }
        return out;
    }

    /**
	 * Supprime les items d'une commande.
	 * @param orderId ID de la commande
	 * @throws SQLException En cas d'erreur SQL
	 */
	public void deleteByOrder(int orderId) throws SQLException {
        try (PreparedStatement ps = db.prepareStatement("DELETE FROM order_items WHERE order_id=?")) {
            ps.setInt(1, orderId);
            ps.executeUpdate();
        }
    }
}
